src/main/java/io/wgj/ctci/Stack.java

```diff
@@ -1,21 +1,80 @@
 package io.wgj.ctci;
 
-public class Stack {
-    int[] stack;
+/**
+ * Stack is a stack data structure backed by a dynamic array. Stack can be
+ * initialized with a starting array size. It's written to develop familiarity
+ * with the stack data structure, and should otherwise not be used.
+ * Maximum allowed array is 1024. If you set a starting capacity larger than
+ * this, the maximum allowed array size is used as the starting capacity. An
+ * UnsupportedOperationException will be thrown when more the maximum allowed
+ * amount of elements are added.
+ */
+public class Stack<E> {
+    static final int MAX_ARRAY_SIZE = 1024;
+    static final int DEFAULT_CAPACITY = 10;
+    Object[] stack;
+    int size;
     int position;
 
-    public Stack(){
-        stack = new int[10];
+    public Stack() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public Stack(int size) {
+        stack = new Object[size];
+        if (size > MAX_ARRAY_SIZE) {
+            size = MAX_ARRAY_SIZE;
+        }
+        this.size = size;
         position = 0;
     }
 
-    public void push(int i) {
-        stack[position] = i;
+    public void push(E e) {
+        if (size == position) {
+            this.grow();
+        }
+        stack[position] = e;
         position++;
     }
 
-    public int pop(){
+    public E pop() {
         position--;
-        return stack[position];
+        @SuppressWarnings("unchecked")
+        E e = (E) stack[position];
+        return e;
+    }
+
+    private void grow() {
+        int newSize = size << 2;
+        if (newSize > MAX_ARRAY_SIZE) {
+            int leftOver = MAX_ARRAY_SIZE - size;
+            // Squeeze out the remaining space we'd lose.
+            if (leftOver > 0) {
+                newSize = leftOver + size;
+            } else {
+                throw new UnsupportedOperationException();
+            }
+        }
```

Comment:
It took me a couple reads through the block above is doing.  I think its equivalent to having:
```java
int newSize;
if (size == MAX_ARRAY_SIZE) {
  // We're out of room to grow.
  throw new UnsupportedOperationException();
} else if (size < MAX_ARRAY_SIZE / 2) {
  // There is enough space to grow the array
  newSize = size * 2;
} else {
  // Doubling the array would be too big, so cap it at the max size.
  newSize = MAX_ARRAY_SIZE;
}
```

If you can pull it off, having fewer nested scopes is more understandable because of how our brains work.  The cognitive load for following branches within branches is much higher than a flat structure.  For example:
```java
if (a) {
  if (b) {
    // Do thing
  } else {
    // Do other thing
  }
}
```
Is equivalent to but less understandable than:
```java
if (a && b) {
  // do thing
} else {
 // Do other thing
}
```

Some static analysis tools even use nested scopes as a measure of quality, and I can't say that I disagree.  Obviously you can't always simplify like this.  Sometimes you have to rejigger the variables around to do it, butsometimes it just isn't possible to express the logic that way.

Reply:
> Some static analysis tools even use nested scopes as a measure of quality.

That's one of the first things we'd show off in a JTest demo, and ironically I know "McCabe's Cyclomatic Complexity" by name. Shame on me.